src/esw/cameras.py

```diff
@@ -0,0 +1,351 @@
+                self.resolution = args
+            except Exception:
+                rospy.logerr(f"Failed to create video source for device {self.device}.")
+                self.video_source = None
+                self.output_by_endpoint = {}
+
+    def is_streaming(self) -> bool:
+        """
+        Returns if streaming or not
+        :return: a bool if streaming
+        """
+        is_streaming = len(self.output_by_endpoint) != 0
+        if is_streaming and self.video_source is not None:
+            return True
+        elif not is_streaming and self.video_source is None:
+            return False
+        else:
+            rospy.logerr("Should not have entered this state.")
+            assert False
+
+
+class StreamingManager:
+
+    _services: List[List[CameraCmd]]
+    _device_lock: threading.Lock
+    _resolution_args: List[List[str]]
+    _endpoints: List[str]
+    _service_streams_by_endpoints: Dict[str, Tuple[int, int]]
+    _video_devices: List[VideoDevices]
+    _active_devices: int
+    _max_devices: int
+
+    def __init__(self):
+
+        self._services = [
+            [CameraCmd(-1, 0), CameraCmd(-1, 0), CameraCmd(-1, 0), CameraCmd(-1, 0)],
+            [CameraCmd(-1, 0), CameraCmd(-1, 0), CameraCmd(-1, 0), CameraCmd(-1, 0)],
+        ]
+        self._video_devices = []
+        for i in range(rospy.get_param("cameras/max_video_device_id_number")):
+            skip_every_other_device = rospy.get_param("cameras/skip_every_other_device")
+            if skip_every_other_device:
+                self._video_devices.append(VideoDevices(i * 2))
+            else:
+                self._video_devices.append(VideoDevices(i))
+        self._endpoints = [
+            rospy.get_param("cameras/endpoints/primary_0"),
+            rospy.get_param("cameras/endpoints/primary_1"),
+            rospy.get_param("cameras/endpoints/primary_2"),
+            rospy.get_param("cameras/endpoints/primary_3"),
+            rospy.get_param("cameras/endpoints/secondary_0"),
+            rospy.get_param("cameras/endpoints/secondary_1"),
+            rospy.get_param("cameras/endpoints/secondary_2"),
+            rospy.get_param("cameras/endpoints/secondary_3"),
+        ]
+        self._service_streams_by_endpoints = {
+            self._endpoints[0]: (0, 0),
+            self._endpoints[1]: (0, 1),
+            self._endpoints[2]: (0, 2),
+            self._endpoints[3]: (0, 3),
+            self._endpoints[4]: (1, 0),
+            self._endpoints[5]: (1, 1),
+            self._endpoints[6]: (1, 2),
+            self._endpoints[7]: (1, 3),
+        }
+        self._resolution_args = [
+            list(rospy.get_param("cameras/arguments/144_res")),
+            list(rospy.get_param("cameras/arguments/360_res")),
+            list(rospy.get_param("cameras/arguments/720_res")),
+        ]
+        self._active_devices = 0
+        # determined by hardware. this is a constant. do not change.
+        self._max_devices = 4
```

Comment:
Should change to all-caps and make static (by placing above the constructor)

Reply:
done